Cache/IO.lean

```diff
@@ -448,15 +454,27 @@ def leanModulesFromSpec (sp : SearchPath) (argₛ : String) :
       -- does not correspond to a Lean file, but to an existing folder
       -- `Mathlib/Data/`
       let folder := sourceFile.withExtension ""
-      IO.println s!"Searching directory {folder} for .lean files"
       if ← folder.pathExists then
-        -- (2.) provided "module name" of an existing folder: walk dir
-        -- TODO: will be implemented in #21838
-        return .error "Entering a part of a module name \
-          (i.e. `Mathlib.Data` when only the folder `Mathlib/Data/` but no \
-          file `Mathlib/Data.lean` exists) is not supported yet!"
+        -- (2.) provided "module name" of an existing folder: walk it
+        IO.println s!"Searching directory {folder} for .lean files"
+        let leanModulesInFolder ← walkDir folder srcDir
+        return .ok leanModulesInFolder
       else
         return .error "Invalid argument: non-existing module {mod}"
+where
+  /-- assumes the folder exists -/
+  walkDir (folder : FilePath) (srcDir : FilePath) : IO <| Array (Name × FilePath) := do
```

Comment:
Does [`Lean.forEachModuleInDir`](https://leanprover-community.github.io/mathlib4_docs/Lean/Util/Path.html#Lean.forEachModuleInDir) help here?

Reply:
Ah, I guess not quite because you want both the name and the filepath.